apiserver/facades/client/bundle/bundle.go

```diff
@@ -542,7 +542,7 @@ func (b *BundleAPI) bundleDataApplications(
 			// from the charm config metadata.
 			cfgInfo, ok := charmConfigCache[charmURL]
 			if !ok {
-				ch, err := backend.Charm(curl)
+				ch, err := backend.Charm(application.CharmURL())
```

Comment:
I don't think application.CharmURL does call into state? I think this could be:
https://github.com/juju/juju/blob/3.3/core/cache/application.go#L69-L71

However, this call here is on an interface, so i think you're right in principle, it probably is good practise to store the result as we don't know how costly this is